packages/core/src/multiCompilerHost.ts

```diff
@@ -253,19 +177,40 @@ export function createMultiCompilerHost(fs: Package): MultiCompilerHost {
     };
   }
 
-  function createTraceCollector() {
-    const traces: string[] = [];
-    return {
-      trace: (message: string) => traces.push(message),
-      read: () => {
-        const result = traces.slice();
-        clear();
-        return result;
-      },
-      clear,
-    };
-    function clear() {
-      traces.length = 0;
-    }
+  private getImpliedNodeFormatForFile(fileName: string): ts.ModuleKind.ESNext | ts.ModuleKind.CommonJS | undefined {
+    return ts.getImpliedNodeFormatForFile(
+      toPath(fileName),
+      this.normalModuleResolutionCache.getPackageJsonInfoCache(),
+      this.compilerHost,
+      this.compilerOptions
+    );
+  }
+
+  private getPackageScopeForPath(fileName: string): ts.PackageJsonInfo | undefined {
+    return ts.getPackageScopeForPath(
+      fileName,
+      ts.getTemporaryModuleResolutionState(
+        // TODO: consider always using the node16 cache because package.json should be a hit
+        this.normalModuleResolutionCache.getPackageJsonInfoCache(),
+        this.compilerHost,
+        this.compilerOptions
+      )
+    );
+  }
+}
+
+class TraceCollector {
+  private traces: string[] = [];
+
+  trace = (message: string) => {
+    this.traces.push(message);
+  };
```

Comment:
I honestly didn't know that you could use an arrow function like this in a class to avoid having to `bind` the function outside but it works.

Reply:
Methods are declared on the class prototype so they can be shared between all instances. This is a class field, which is initialized on each instance during construction (equivalent to doing `Object.defineProperty(this, "trace", ...)` in the constructor). This has a perf/memory cost as compared to a true method, but of course, the cost is equivalent to every other way of accomplishing what we’re trying to accomplish here—`.bind` also creates a new copy of the function; it was just happening at a different time and place. There’s no way around creating a new function identity for each `CompilerHost["trace"]`. (As a piece of trivia, this was a common trick for React devs in the days of class components for event handlers, so you could do `<button onClick={this.onClick}>` instead of `this.onClick.bind(this)`.) This still has a bit of a weird smell to me since you have to recognize which functions need this weird syntactic trick and which don’t. On the upside though, it’s a bit more efficient since at least `read` and `clear` are shared between all instances. It’s mostly a personal preference. I don’t love it but I appreciate the contribution so I’ll merge it 😛